torchrl/envs/libs/unity.py

```diff
@@ -0,0 +1,421 @@
+        ...         side_channels=[],
+        ...         additional_args=[],
+        ...         log_folder=<<PATH TO WHERE LOGS SHOULD BE STORED>>,
+        ...         device=device,
+        ...     )
+        ... )
+    """
+
+    git_url = "https://github.com/Unity-Technologies/ml-agents"
+    libname = "mlagents_envs"
+
+    def __init__(self, env=None, **kwargs):
+        if env is not None:
+            kwargs["env"] = env
+        super().__init__(**kwargs)
+
+    def _init_env(self):
+        pass
+
+    def _compute_num_agents(self, env):
+        num_agents = 0
+        for behavior_name in env.behavior_specs.keys():
+            decision_steps, terminal_steps = env.get_steps(behavior_name)
+            num_agents += len(decision_steps) + len(terminal_steps)
+        return num_agents
+
+    def _set_seed(self, seed: int | None):
+        warn(
+            "Seeding through _set_seed has not been implemented. Please set the "
+            "seed when you create the environment."
+        )
+
+    @_classproperty
+    def available_envs(cls) -> list[str]:
+        return []
+
+    def _build_env(self, env: BaseEnv):
+        if not env.behavior_specs:
+            # Take a single step so that the brain information will be sent over
+            env.step()
+        self._behavior_names = list(env.behavior_specs.keys())
+        self.num_agents = self._compute_num_agents(env)
+        self._agent_id_to_behavior = {}
+        return env
+
+    def _make_specs(self, env: BaseEnv) -> None:
+        observation_specs = [None] * self.num_agents
+        behavior_id_specs = [None] * self.num_agents
+        agent_id_specs = [None] * self.num_agents
+        action_specs = [None] * self.num_agents
+        reward_specs = [None] * self.num_agents
+        done_specs = [None] * self.num_agents
+        valid_mask_specs = [None] * self.num_agents
+
+        for behavior_name, behavior_unity_spec in env.behavior_specs.items():
+            decision_steps, terminal_steps = env.get_steps(behavior_name)
+            for steps in [decision_steps, terminal_steps]:
+                for agent_id in steps.agent_id:
+                    self._agent_id_to_behavior[agent_id] = behavior_name
+
+                    observation_specs[agent_id] = CompositeSpec(
+                        {
+                            f"obs_{i}": _unity_to_torchrl_spec_transform(
+                                spec, dtype=np.dtype("float32"), device=self.device
+                            )
+                            for i, spec in enumerate(
+                                behavior_unity_spec.observation_specs
+                            )
+                        }
+                    )
+                    behavior_id_specs[agent_id] = UnboundedDiscreteTensorSpec(
+                        shape=1, device=self.device, dtype=torch.int8
+                    )
+                    agent_id_specs[agent_id] = UnboundedDiscreteTensorSpec(
```

Comment:
Can you elaborate a bit more on why we have an agent_is_specs and a behavior_id_spec.

Is this data gonna chage at each step?

Reply:
No, it won't. However, I wasn't sure if it might be useful for users to have access to this information. I don't have a strong opinion on this, and we can remove it if we don't think it's useful.

The behavior_id might be more important especially since we don't have the information grouped by behaviors yet. (Option 3 in our discussion from earlier).

Reply:
we could have an attribute in the env object.

we do not need to put it in every tensordict or in the specs

Reply:
Just updated this!